In [84]:
import pandas as pd
import numpy as np
from nltk.tokenize import word_tokenize
from nltk.corpus import stopwords
from nltk import WordNetLemmatizer
from nltk.corpus import wordnet
from nltk import pos_tag
from sklearn.preprocessing import LabelEncoder
from sklearn.base import BaseEstimator, TransformerMixin
from imblearn.pipeline import Pipeline
from imblearn.over_sampling import SMOTE
from sklearn import set_config
set_config(display="diagram")
from sklearn.naive_bayes import ComplementNB
from sklearn.metrics import classification_report
from sklearn.compose import ColumnTransformer
from sklearn.metrics import plot_roc_curve, roc_auc_score,plot_confusion_matrix
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer, CountVectorizer
from sklearn.model_selection import GridSearchCV
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.metrics import confusion_matrix, ConfusionMatrixDisplay
from sklearn.metrics import accuracy_score, classification_report
from sklearn.metrics import plot_roc_curve, auc,roc_curve,roc_auc_score
from sklearn.naive_bayes import MultinomialNB
from sklearn.svm import SVC

In [85]:
df = pd.read_csv('/Users/mahmoud/Car_Price_Prediction/complete_car_data.csv')

In [86]:
df

Car Name      Mileage             Dealer Name  \
0              2016 Nissan Versa 1.6 S+   60,001 mi.         Bical Auto Mall   
1                      1989 Mazda RX-7    90,462 mi.       Brick City Motors   
2             2011 Chrysler 300 Limited  129,270 mi.   Dash Auto Gallery Inc   
3                 2016 GMC Acadia SLT-1  146,121 mi.   Dash Auto Gallery Inc   
4                    2006 Lexus RX 330   129,434 mi.  Big T's Auto Sales LLC   
...                                 ...          ...                     ...   
7253  2011 Subaru Forester 2.5X Limited  130,848 mi.               Motors 46   
7254          2011 Hyundai Santa Fe GLS  124,413 mi.               Motors 46   
7255                     2018 Kia Rio S   61,380 mi.           Napoli Nissan   
7256            2016 GMC Terrain Denali  105,678 mi.  Lester Glenn Chevrolet   
7257           2016 Nissan Pathfinder S  115,472 mi.        J&S Autohaus III   

      Rating   Review Count    Price Car Rating  \
0        4.7  (746 reviews)  $10,699        5.0   
1        3.0   (85 reviews)   $7,995        3.0   
2        4.8   (81 reviews)   $9,891        4.8   
3        4.8   (81 reviews)  $14,291        4.8   
4        NaN    (5 reviews)   $8,795        4.5   
...      ...            ...      ...        ...   
7253     NaN            NaN   $7,499        4.3   
7254     NaN            NaN   $6,995        4.5   
7255     NaN            NaN  $14,800        5.0   
7256     NaN            NaN  $14,447        4.9   
7257     NaN            NaN  $12,995        4.2   

                                            Car Reviews  \
0     ['April 10, 2023\nBy Martin W from Syracuse, N...   
1                                                    []   
2     ['February 11, 2023\nBy Asa from Redding CA \n...   
3     ['January 30, 2023\nBy Paul  from Highlands Ra...   
4     ['September  1, 2022\nBy Jeff from Ft. Collins...   
...                                                 ...   
7253  ['October  3, 2022\nBy KD from Fairhope, Alaba...   
7254  ['March 10, 2023\nBy notme96 from Spokane, WA\...   
7255  ['April 12, 2023\nBy Brenda Shrader  from Kans...   
7256  ['May  3, 2021\nBy Traci P from Anadarko\nOwns...   
7257  ['December 15, 2022\nBy Richard Horvath  from ...   

                         Accidents or damage 1-owner vehicle  ...  \
0     At least 1 accident or damage reported              No  ...   
1                              None reported              No  ...   
2     At least 1 accident or damage reported              No  ...   
3                              None reported              No  ...   
4                              None reported              No  ...   
...                                      ...             ...  ...   
7253  At least 1 accident or damage reported              No  ...   
7254                           None reported              No  ...   
7255                           None reported             Yes  ...   
7256                           None reported             Yes  ...   
7257  At least 1 accident or damage reported              No  ...   

                          Open recall       Exterior color  \
0                       Not available                Black   
1                       Not available                Black   
2                       Not available                 Blue   
3     At least 1 open recall reported                White   
4                       Not available                White   
...                               ...                  ...   
7253                    Not available  Sage Green Metallic   
7254  At least 1 open recall reported         Mineral Gray   
7255                    Not available          Clear White   
7256                    Not available  White Frost Tricoat   
7257  At least 1 open recall reported        Glacier White   

           Interior color         Drivetrain  \
0                Charcoal  Front-wheel Drive   
1                   Black   Rear-wheel Drive   
2     Black / Light Frost   Re

In [87]:
# Having the mileage column in float rather than string
df['Mileage'] = df['Mileage'].str.replace(' mi.', '').str.replace(',', '')
df['Mileage'] = pd.to_numeric(df['Mileage'], errors='coerce')
df['Mileage']

0        60001.0
1        90462.0
2       129270.0
3       146121.0
4       129434.0
          ...   
7253    130848.0
7254    124413.0
7255     61380.0
7256    105678.0
7257    115472.0
Name: Mileage, Length: 7258, dtype: float64

In [88]:
df[['Year', 'Make', 'Model', 'Type']] = df['Car Name'].str.extract(r'(\d{4})\s+([^ ]+)\s+([^ ]+)\s+(.*)')
df

Car Name   Mileage             Dealer Name  \
0              2016 Nissan Versa 1.6 S+   60001.0         Bical Auto Mall   
1                      1989 Mazda RX-7    90462.0       Brick City Motors   
2             2011 Chrysler 300 Limited  129270.0   Dash Auto Gallery Inc   
3                 2016 GMC Acadia SLT-1  146121.0   Dash Auto Gallery Inc   
4                    2006 Lexus RX 330   129434.0  Big T's Auto Sales LLC   
...                                 ...       ...                     ...   
7253  2011 Subaru Forester 2.5X Limited  130848.0               Motors 46   
7254          2011 Hyundai Santa Fe GLS  124413.0               Motors 46   
7255                     2018 Kia Rio S   61380.0           Napoli Nissan   
7256            2016 GMC Terrain Denali  105678.0  Lester Glenn Chevrolet   
7257           2016 Nissan Pathfinder S  115472.0        J&S Autohaus III   

      Rating   Review Count    Price Car Rating  \
0        4.7  (746 reviews)  $10,699        5.0   
1        3.0   (85 reviews)   $7,995        3.0   
2        4.8   (81 reviews)   $9,891        4.8   
3        4.8   (81 reviews)  $14,291        4.8   
4        NaN    (5 reviews)   $8,795        4.5   
...      ...            ...      ...        ...   
7253     NaN            NaN   $7,499        4.3   
7254     NaN            NaN   $6,995        4.5   
7255     NaN            NaN  $14,800        5.0   
7256     NaN            NaN  $14,447        4.9   
7257     NaN            NaN  $12,995        4.2   

                                            Car Reviews  \
0     ['April 10, 2023\nBy Martin W from Syracuse, N...   
1                                                    []   
2     ['February 11, 2023\nBy Asa from Redding CA \n...   
3     ['January 30, 2023\nBy Paul  from Highlands Ra...   
4     ['September  1, 2022\nBy Jeff from Ft. Collins...   
...                                                 ...   
7253  ['October  3, 2022\nBy KD from Fairhope, Alaba...   
7254  ['March 10, 2023\nBy notme96 from Spokane, WA\...   
7255  ['April 12, 2023\nBy Brenda Shrader  from Kans...   
7256  ['May  3, 2021\nBy Traci P from Anadarko\nOwns...   
7257  ['December 15, 2022\nBy Richard Horvath  from ...   

                         Accidents or damage 1-owner vehicle  ...  \
0     At least 1 accident or damage reported              No  ...   
1                              None reported              No  ...   
2     At least 1 accident or damage reported              No  ...   
3                              None reported              No  ...   
4                              None reported              No  ...   
...                                      ...             ...  ...   
7253  At least 1 accident or damage reported              No  ...   
7254                           None reported              No  ...   
7255                           None reported             Yes  ...   
7256                           None reported             Yes  ...   
7257  At least 1 accident or damage reported              No  ...   

                                                    MPG      Fuel type  \
0     31–40\n\n\n\nBased on EPA mileage ratings. Use...       Gasoline   
1                                                     –       Gasoline   
2     18–27\n\n\n\nBased on EPA mileage ratings. Use...  E85 Flex Fuel   
3     15–22\n\n\n\nBased on EPA mileage ratings. Use...       Gasoline   
4     19–25\n\n\n\nBased on EPA mileage ratings. Use...       Gasoline   
...                                                 ...            ...   
7253  21–27\n\n\n\nBased on EPA mileage ratings. Use...       Gasoline   
7254  20–25\n\n\n\nBased on EPA mileage ratings. Use...       Gasoline   
7255  28–37\n\n\n\nBased on EPA mileage ratings. Use...       Gasoline   
7256  16–23\n\n\n\nBased on EPA mileage ratings. Use...       Gasoline   
7257  19–26\n\n\n\nBased on EPA mileage ratings. Use...       Gasoline   

           Transmission                                           

In [89]:
df.isna().sum()

Car Name                  0
Mileage                   5
Dealer Name             189
Rating                 4622
Review Count           3215
Price                     0
Car Rating                0
Car Reviews               0
Accidents or damage       0
1-owner vehicle           0
Personal use only         0
Open recall               0
Exterior color            0
Interior color            0
Drivetrain                0
MPG                     195
Fuel type                 8
Transmission              0
Engine                    0
VIN                       0
Stock #                 189
Year                      0
Make                      0
Model                     0
Type                      0
dtype: int64

# Nans:
- Mileage                   5
- Dealer Name             189
- Rating                 4622
- Review Count           3215
- MPG                     195
- Stock #                 189
- Fuel type              8

In [90]:
df = df.rename(columns={"Rating": "Dealer Rating"})

In [91]:
# Replacing Dealer Name with Unkown
df['Dealer Name'] = df['Dealer Name'].fillna('Unkown')

In [92]:
df[df['Review Count'].isna()]

Car Name   Mileage  \
42              2007 INFINITI G35 Sport  122384.0   
74             2017 Hyundai Santa Fe SE   93416.0   
76          2004 Cadillac Escalade Base  127381.0   
77                    2008 Volvo C70 T5  132242.0   
90    2011 Mercedes-Benz S-Class 4MATIC  115492.0   
...                                 ...       ...   
7253  2011 Subaru Forester 2.5X Limited  130848.0   
7254          2011 Hyundai Santa Fe GLS  124413.0   
7255                     2018 Kia Rio S   61380.0   
7256            2016 GMC Terrain Denali  105678.0   
7257           2016 Nissan Pathfinder S  115472.0   

                       Dealer Name  Dealer Rating Review Count    Price  \
42                          Unkown            NaN          NaN  $14,200   
74            Price Wise Auto Sale            NaN          NaN  $12,995   
76                         Carz4Us            NaN          NaN   $6,995   
77                         Carz4Us            NaN          NaN   $7,395   
90    Certified Cars of Huntington            NaN          NaN  $11,595   
...                            ...            ...          ...      ...   
7253                     Motors 46            NaN          NaN   $7,499   
7254                     Motors 46            NaN          NaN   $6,995   
7255                 Napoli Nissan            NaN          NaN  $14,800   
7256        Lester Glenn Chevrolet            NaN          NaN  $14,447   
7257              J&S Autohaus III            NaN          NaN  $12,995   

     Car Rating                                        Car Reviews  \
42          4.7  ['August 30, 2020\nBy alberto from DALLAS\nOwn...   
74          4.8  ['January 20, 2023\nBy Eddie from Chicago, IL\...   
76          4.5  ['May 14, 2021\nBy Lefty69 from Hancock, Maryl...   
77          4.7  ['December 14, 2019\nBy Jim from Charlotte, NC...   
90          4.8  ['March 25, 2023\nBy david v from california \...   
...         ...                                                ...   
7253        4.3  ['October  3, 2022\nBy KD from Fairhope, Alaba...   
7254        4.5  ['March 10, 2023\nBy notme96 from Spokane, WA\...   
7255        5.0  ['April 12, 2023\nBy Brenda Shrader  from Kans...   
7256        4.9  ['May  3, 2021\nBy Traci P from Anadarko\nOwns...   
7257        4.2  ['December 15, 2022\nBy Richard Horvath  from ...   

                         Accidents or damage 1-owner vehicle  ...  \
42                             None reported              No  ...   
74    At least 1 accident or damage reported             Yes  ...   
76                             None reported              No  ...   
77                             None reported             Yes  ...   
90                             None reported              No  ...   
...                                      ...             ...  ...   
7253  At least 1 accident or damage reported              No  ...   
7254                           None reported              No  ...   
7255                           None reported             Yes  ...   
7256                           None reported             Yes  ...   
7257  At least 1 accident or damage reported              No  ...   

                                                    MPG Fuel type  \
42                                                  NaN  Gasoline   
74    18–25\n\n\n\nBased on EPA mileage ratings. Use...  Gasoline   
76    13–17\n\n\n\nBased on EPA mileage ratings. Use...  Gasoline   
77    18–27\n\n\n\nBased on EPA mileage ratings. Use...  Gasoline   
90    14–21\n\n\n\nBased on EPA mileage ratings. Use...  Gasoline   
...                                                 ...       ...   
7253  21–27\n\n\n\nBased on EPA mileage ratings. Use...  Gasoline   
7254  20–25\n\n\n\nBased on EPA mileage ratings. Use...  Gasoline   
7255  28–37\n\n\n\nBased on EPA mileage ratings. Use...  Gasoline   
7256  16–23\n\n\n\nBased on EPA mileage ratings. Use...  Gasoline   
7257  19–26\n\n\n\nBased on EPA mileage ratings. Use...  Gasolin

In [93]:
# All Nans in rating to the mean
rating_mean = df['Dealer Rating'].mean()
df['Dealer Rating'].fillna(rating_mean, inplace=True)

In [94]:
# So we can fill those counts with 0 as well
df['Review Count'] = df['Review Count'].fillna(0)

In [95]:
df.isna().sum()

Car Name                 0
Mileage                  5
Dealer Name              0
Dealer Rating            0
Review Count             0
Price                    0
Car Rating               0
Car Reviews              0
Accidents or damage      0
1-owner vehicle          0
Personal use only        0
Open recall              0
Exterior color           0
Interior color           0
Drivetrain               0
MPG                    195
Fuel type                8
Transmission             0
Engine                   0
VIN                      0
Stock #                189
Year                     0
Make                     0
Model                    0
Type                     0
dtype: int64

In [96]:
# NO need for stock # so get rid of it
df.drop(['Stock #'], axis =1, inplace=True)

In [97]:
# dropping the mileage Nans
df = df.dropna(subset=['Mileage'])

In [98]:
df

Car Name   Mileage             Dealer Name  \
0              2016 Nissan Versa 1.6 S+   60001.0         Bical Auto Mall   
1                      1989 Mazda RX-7    90462.0       Brick City Motors   
2             2011 Chrysler 300 Limited  129270.0   Dash Auto Gallery Inc   
3                 2016 GMC Acadia SLT-1  146121.0   Dash Auto Gallery Inc   
4                    2006 Lexus RX 330   129434.0  Big T's Auto Sales LLC   
...                                 ...       ...                     ...   
7253  2011 Subaru Forester 2.5X Limited  130848.0               Motors 46   
7254          2011 Hyundai Santa Fe GLS  124413.0               Motors 46   
7255                     2018 Kia Rio S   61380.0           Napoli Nissan   
7256            2016 GMC Terrain Denali  105678.0  Lester Glenn Chevrolet   
7257           2016 Nissan Pathfinder S  115472.0        J&S Autohaus III   

      Dealer Rating   Review Count    Price Car Rating  \
0            4.7000  (746 reviews)  $10,699        5.0   
1            3.0000   (85 reviews)   $7,995        3.0   
2            4.8000   (81 reviews)   $9,891        4.8   
3            4.8000   (81 reviews)  $14,291        4.8   
4            3.5272    (5 reviews)   $8,795        4.5   
...             ...            ...      ...        ...   
7253         3.5272              0   $7,499        4.3   
7254         3.5272              0   $6,995        4.5   
7255         3.5272              0  $14,800        5.0   
7256         3.5272              0  $14,447        4.9   
7257         3.5272              0  $12,995        4.2   

                                            Car Reviews  \
0     ['April 10, 2023\nBy Martin W from Syracuse, N...   
1                                                    []   
2     ['February 11, 2023\nBy Asa from Redding CA \n...   
3     ['January 30, 2023\nBy Paul  from Highlands Ra...   
4     ['September  1, 2022\nBy Jeff from Ft. Collins...   
...                                                 ...   
7253  ['October  3, 2022\nBy KD from Fairhope, Alaba...   
7254  ['March 10, 2023\nBy notme96 from Spokane, WA\...   
7255  ['April 12, 2023\nBy Brenda Shrader  from Kans...   
7256  ['May  3, 2021\nBy Traci P from Anadarko\nOwns...   
7257  ['December 15, 2022\nBy Richard Horvath  from ...   

                         Accidents or damage 1-owner vehicle  ...  \
0     At least 1 accident or damage reported              No  ...   
1                              None reported              No  ...   
2     At least 1 accident or damage reported              No  ...   
3                              None reported              No  ...   
4                              None reported              No  ...   
...                                      ...             ...  ...   
7253  At least 1 accident or damage reported              No  ...   
7254                           None reported              No  ...   
7255                           None reported             Yes  ...   
7256                           None reported             Yes  ...   
7257  At least 1 accident or damage reported              No  ...   

             Drivetrain                                                MPG  \
0     Front-wheel Drive  31–40\n\n\n\nBased on EPA mileage ratings. Use...   
1      Rear-wheel Drive                                                  –   
2      Rear-wheel Drive  18–27\n\n\n\nBased on EPA mileage ratings. Use...   
3       All-wheel Drive  15–22\n\n\n\nBased on EPA mileage ratings. Use...   
4     Front-wheel Drive  19–25\n\n\n\nBased on EPA mileage ratings. Use...   
...                 ...                                                ...   
7253    All-wheel Drive  21–27\n\n\n\nBased on EPA mileage ratings. Use...   
7254    All-wheel Drive  20–25\n\n\n\nBased on EPA mileage ratings. Use...   
7255  Front-wheel Drive  28–37\n\n\n\nBased on EPA mileage ratings. Use...   
7256    All-wheel Drive  16–23\n\n\n\nBased on EPA mileage ratings. Use...   
7257   Four-w

In [99]:
# we can fill mpg with the average mpg which is 25.4
df['MPG'] = df['MPG'].fillna(25.4)

In [100]:
# we can fill fuel type with gasoline
df['Fuel type'] = df['Fuel type'].fillna('Gasoline')

In [101]:
df

Car Name   Mileage             Dealer Name  \
0              2016 Nissan Versa 1.6 S+   60001.0         Bical Auto Mall   
1                      1989 Mazda RX-7    90462.0       Brick City Motors   
2             2011 Chrysler 300 Limited  129270.0   Dash Auto Gallery Inc   
3                 2016 GMC Acadia SLT-1  146121.0   Dash Auto Gallery Inc   
4                    2006 Lexus RX 330   129434.0  Big T's Auto Sales LLC   
...                                 ...       ...                     ...   
7253  2011 Subaru Forester 2.5X Limited  130848.0               Motors 46   
7254          2011 Hyundai Santa Fe GLS  124413.0               Motors 46   
7255                     2018 Kia Rio S   61380.0           Napoli Nissan   
7256            2016 GMC Terrain Denali  105678.0  Lester Glenn Chevrolet   
7257           2016 Nissan Pathfinder S  115472.0        J&S Autohaus III   

      Dealer Rating   Review Count    Price Car Rating  \
0            4.7000  (746 reviews)  $10,699        5.0   
1            3.0000   (85 reviews)   $7,995        3.0   
2            4.8000   (81 reviews)   $9,891        4.8   
3            4.8000   (81 reviews)  $14,291        4.8   
4            3.5272    (5 reviews)   $8,795        4.5   
...             ...            ...      ...        ...   
7253         3.5272              0   $7,499        4.3   
7254         3.5272              0   $6,995        4.5   
7255         3.5272              0  $14,800        5.0   
7256         3.5272              0  $14,447        4.9   
7257         3.5272              0  $12,995        4.2   

                                            Car Reviews  \
0     ['April 10, 2023\nBy Martin W from Syracuse, N...   
1                                                    []   
2     ['February 11, 2023\nBy Asa from Redding CA \n...   
3     ['January 30, 2023\nBy Paul  from Highlands Ra...   
4     ['September  1, 2022\nBy Jeff from Ft. Collins...   
...                                                 ...   
7253  ['October  3, 2022\nBy KD from Fairhope, Alaba...   
7254  ['March 10, 2023\nBy notme96 from Spokane, WA\...   
7255  ['April 12, 2023\nBy Brenda Shrader  from Kans...   
7256  ['May  3, 2021\nBy Traci P from Anadarko\nOwns...   
7257  ['December 15, 2022\nBy Richard Horvath  from ...   

                         Accidents or damage 1-owner vehicle  ...  \
0     At least 1 accident or damage reported              No  ...   
1                              None reported              No  ...   
2     At least 1 accident or damage reported              No  ...   
3                              None reported              No  ...   
4                              None reported              No  ...   
...                                      ...             ...  ...   
7253  At least 1 accident or damage reported              No  ...   
7254                           None reported              No  ...   
7255                           None reported             Yes  ...   
7256                           None reported             Yes  ...   
7257  At least 1 accident or damage reported              No  ...   

             Drivetrain                                                MPG  \
0     Front-wheel Drive  31–40\n\n\n\nBased on EPA mileage ratings. Use...   
1      Rear-wheel Drive                                                  –   
2      Rear-wheel Drive  18–27\n\n\n\nBased on EPA mileage ratings. Use...   
3       All-wheel Drive  15–22\n\n\n\nBased on EPA mileage ratings. Use...   
4     Front-wheel Drive  19–25\n\n\n\nBased on EPA mileage ratings. Use...   
...                 ...                                                ...   
7253    All-wheel Drive  21–27\n\n\n\nBased on EPA mileage ratings. Use...   
7254    All-wheel Drive  20–25\n\n\n\nBased on EPA mileage ratings. Use...   
7255  Front-wheel Drive  28–37\n\n\n\nBased on EPA mileage ratings. Use...   
7256    All-wheel Drive  16–23\n\n\n\nBased on EPA mileage ratings. Use...   
7257   Four-w

In [102]:
df.isna().sum()

Car Name               0
Mileage                0
Dealer Name            0
Dealer Rating          0
Review Count           0
Price                  0
Car Rating             0
Car Reviews            0
Accidents or damage    0
1-owner vehicle        0
Personal use only      0
Open recall            0
Exterior color         0
Interior color         0
Drivetrain             0
MPG                    0
Fuel type              0
Transmission           0
Engine                 0
VIN                    0
Year                   0
Make                   0
Model                  0
Type                   0
dtype: int64

# NO more nulls

In [103]:
# column_data_types = df.dtypes

# print(column_data_types)
df.dtypes

Car Name                object
Mileage                float64
Dealer Name             object
Dealer Rating          float64
Review Count            object
Price                   object
Car Rating              object
Car Reviews             object
Accidents or damage     object
1-owner vehicle         object
Personal use only       object
Open recall             object
Exterior color          object
Interior color          object
Drivetrain              object
MPG                     object
Fuel type               object
Transmission            object
Engine                  object
VIN                     object
Year                    object
Make                    object
Model                   object
Type                    object
dtype: object

In [104]:
# Since Price is a string, lets convert to a float

# First remove dollar sign
df['Price'] = df['Price'].str.replace('[^\d.]', '', regex=True)

# Now make it a float
df['Price'] = df['Price'].astype(float)

# Confirm the updated data type
print(df['Price'].dtype)

float64


In [105]:
df['Car Rating'] = pd.to_numeric(df['Car Rating'], errors='coerce')
mean_car_rating = df.loc[df['Car Rating'] != 'No Rating Available', 'Car Rating'].mean()
df['Car Rating'] = df['Car Rating'].replace('No Rating Available', mean_car_rating)
df['Car Rating'] = df['Car Rating'].astype(float)

In [106]:
df['Review Count'] = df['Review Count'].str.extract('(\d+)').astype(float)
df.dtypes

Car Name                object
Mileage                float64
Dealer Name             object
Dealer Rating          float64
Review Count           float64
Price                  float64
Car Rating             float64
Car Reviews             object
Accidents or damage     object
1-owner vehicle         object
Personal use only       object
Open recall             object
Exterior color          object
Interior color          object
Drivetrain              object
MPG                     object
Fuel type               object
Transmission            object
Engine                  object
VIN                     object
Year                    object
Make                    object
Model                   object
Type                    object
dtype: object

In [107]:
df['Open recall']

0                         Not available
1                         Not available
2                         Not available
3       At least 1 open recall reported
4                         Not available
                     ...               
7253                      Not available
7254    At least 1 open recall reported
7255                      Not available
7256                      Not available
7257    At least 1 open recall reported
Name: Open recall, Length: 7253, dtype: object

In [108]:
df

Car Name   Mileage             Dealer Name  \
0              2016 Nissan Versa 1.6 S+   60001.0         Bical Auto Mall   
1                      1989 Mazda RX-7    90462.0       Brick City Motors   
2             2011 Chrysler 300 Limited  129270.0   Dash Auto Gallery Inc   
3                 2016 GMC Acadia SLT-1  146121.0   Dash Auto Gallery Inc   
4                    2006 Lexus RX 330   129434.0  Big T's Auto Sales LLC   
...                                 ...       ...                     ...   
7253  2011 Subaru Forester 2.5X Limited  130848.0               Motors 46   
7254          2011 Hyundai Santa Fe GLS  124413.0               Motors 46   
7255                     2018 Kia Rio S   61380.0           Napoli Nissan   
7256            2016 GMC Terrain Denali  105678.0  Lester Glenn Chevrolet   
7257           2016 Nissan Pathfinder S  115472.0        J&S Autohaus III   

      Dealer Rating  Review Count    Price  Car Rating  \
0            4.7000         746.0  10699.0         5.0   
1            3.0000          85.0   7995.0         3.0   
2            4.8000          81.0   9891.0         4.8   
3            4.8000          81.0  14291.0         4.8   
4            3.5272           5.0   8795.0         4.5   
...             ...           ...      ...         ...   
7253         3.5272           NaN   7499.0         4.3   
7254         3.5272           NaN   6995.0         4.5   
7255         3.5272           NaN  14800.0         5.0   
7256         3.5272           NaN  14447.0         4.9   
7257         3.5272           NaN  12995.0         4.2   

                                            Car Reviews  \
0     ['April 10, 2023\nBy Martin W from Syracuse, N...   
1                                                    []   
2     ['February 11, 2023\nBy Asa from Redding CA \n...   
3     ['January 30, 2023\nBy Paul  from Highlands Ra...   
4     ['September  1, 2022\nBy Jeff from Ft. Collins...   
...                                                 ...   
7253  ['October  3, 2022\nBy KD from Fairhope, Alaba...   
7254  ['March 10, 2023\nBy notme96 from Spokane, WA\...   
7255  ['April 12, 2023\nBy Brenda Shrader  from Kans...   
7256  ['May  3, 2021\nBy Traci P from Anadarko\nOwns...   
7257  ['December 15, 2022\nBy Richard Horvath  from ...   

                         Accidents or damage 1-owner vehicle  ...  \
0     At least 1 accident or damage reported              No  ...   
1                              None reported              No  ...   
2     At least 1 accident or damage reported              No  ...   
3                              None reported              No  ...   
4                              None reported              No  ...   
...                                      ...             ...  ...   
7253  At least 1 accident or damage reported              No  ...   
7254                           None reported              No  ...   
7255                           None reported             Yes  ...   
7256                           None reported             Yes  ...   
7257  At least 1 accident or damage reported              No  ...   

             Drivetrain                                                MPG  \
0     Front-wheel Drive  31–40\n\n\n\nBased on EPA mileage ratings. Use...   
1      Rear-wheel Drive                                                  –   
2      Rear-wheel Drive  18–27\n\n\n\nBased on EPA mileage ratings. Use...   
3       All-wheel Drive  15–22\n\n\n\nBased on EPA mileage ratings. Use...   
4     Front-wheel Drive  19–25\n\n\n\nBased on EPA mileage ratings. Use...   
...                 ...                                                ...   
7253    All-wheel Drive  21–27\n\n\n\nBased on EPA mileage ratings. Use...   
7254    All-wheel Drive  20–25\n\n\n\nBased on EPA mileage ratings. Use...   
7255  Front-wheel Drive  28–37\n\n\n\nBased on EPA mileage ratings. Use...   
7256    All-wheel Drive  16–23\n\n\n\nBased on EPA mileage ratings. Use...   
7257   Four-w

In [115]:
df

Car Name   Mileage             Dealer Name  \
0              2016 Nissan Versa 1.6 S+   60001.0         Bical Auto Mall   
1                      1989 Mazda RX-7    90462.0       Brick City Motors   
2             2011 Chrysler 300 Limited  129270.0   Dash Auto Gallery Inc   
3                 2016 GMC Acadia SLT-1  146121.0   Dash Auto Gallery Inc   
4                    2006 Lexus RX 330   129434.0  Big T's Auto Sales LLC   
...                                 ...       ...                     ...   
7253  2011 Subaru Forester 2.5X Limited  130848.0               Motors 46   
7254          2011 Hyundai Santa Fe GLS  124413.0               Motors 46   
7255                     2018 Kia Rio S   61380.0           Napoli Nissan   
7256            2016 GMC Terrain Denali  105678.0  Lester Glenn Chevrolet   
7257           2016 Nissan Pathfinder S  115472.0        J&S Autohaus III   

      Dealer Rating  Review Count    Price  Car Rating  \
0            4.7000         746.0  10699.0         5.0   
1            3.0000          85.0   7995.0         3.0   
2            4.8000          81.0   9891.0         4.8   
3            4.8000          81.0  14291.0         4.8   
4            3.5272           5.0   8795.0         4.5   
...             ...           ...      ...         ...   
7253         3.5272           NaN   7499.0         4.3   
7254         3.5272           NaN   6995.0         4.5   
7255         3.5272           NaN  14800.0         5.0   
7256         3.5272           NaN  14447.0         4.9   
7257         3.5272           NaN  12995.0         4.2   

                                            Car Reviews  \
0     ['April 10, 2023\nBy Martin W from Syracuse, N...   
1                                                    []   
2     ['February 11, 2023\nBy Asa from Redding CA \n...   
3     ['January 30, 2023\nBy Paul  from Highlands Ra...   
4     ['September  1, 2022\nBy Jeff from Ft. Collins...   
...                                                 ...   
7253  ['October  3, 2022\nBy KD from Fairhope, Alaba...   
7254  ['March 10, 2023\nBy notme96 from Spokane, WA\...   
7255  ['April 12, 2023\nBy Brenda Shrader  from Kans...   
7256  ['May  3, 2021\nBy Traci P from Anadarko\nOwns...   
7257  ['December 15, 2022\nBy Richard Horvath  from ...   

                         Accidents or damage 1-owner vehicle  ...  \
0     At least 1 accident or damage reported              No  ...   
1                              None reported              No  ...   
2     At least 1 accident or damage reported              No  ...   
3                              None reported              No  ...   
4                              None reported              No  ...   
...                                      ...             ...  ...   
7253  At least 1 accident or damage reported              No  ...   
7254                           None reported              No  ...   
7255                           None reported             Yes  ...   
7256                           None reported             Yes  ...   
7257  At least 1 accident or damage reported              No  ...   

           Interior color         Drivetrain  \
0                Charcoal  Front-wheel Drive   
1                   Black   Rear-wheel Drive   
2     Black / Light Frost   Rear-wheel Drive   
3           Dark Cashmere    All-wheel Drive   
4                     Tan  Front-wheel Drive   
...                   ...                ...   
7253                 Gray    All-wheel Drive   
7254                 Gray    All-wheel Drive   
7255                Black  Front-wheel Drive   
7256            Jet Black    All-wheel Drive   
7257             Charcoal   Four-wheel Drive   

                                                    MPG      Fuel type  \
0     31–40\n\n\n\nBased on EPA mileage ratings. Use...       Gasoline   
1                                                     –       Gasoline   
2     18–27\n\n\n\nBased on EPA mileage ratings. Use...  E85 Flex Fuel  